jctools-experimental/src/main/java/org/jctools/stacks/WFIStack.java

```diff
@@ -0,0 +1,329 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jctools.stacks;
+
+import org.jctools.util.UnsafeAccess;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * A Wait-Free Intrusive Stack (Last-In, First-Out).
+ * <p>
+ * This stack has wait-free push, and wait-free bulk pop and bulk replace operations.
+ * The single-item pop operation is non-blocking but not wait-free.
+ * <p>
+ * The stack supports multiple concurrent consumers and multiple concurrent producers.
+ * <p>
+ * The stack is intrusive, and implementors must extend the {@link Node} class in order to use the stack.
+ * These node objects cannot be reused, and must be allocated fresh for every push.
+ * <p>
+ * The implementation draws inspiration from Dmitry Vyukov's
+ * <a href="http://www.1024cores.net/home/lock-free-algorithms/queues/intrusive-mpsc-node-based-queue">
+ * Intrusive MPSC node-based queue</a>, but is oriented as a stack instead of a queue, and supports MPMC.
+ *
+ * @param <T> The type of node in this stack.
+ */
+@SuppressWarnings({"unchecked", "NullableProblems"})
+public class WFIStack<T extends WFIStack.Node> implements Iterable<T>
+{
+    /**
+     * The super-class of all nodes, or entries, for {@link WFIStack wait-free intrusive stacks}.
+     */
+    public static abstract class Node
+    {
+        volatile Node next;
+        int count;
+
+        final Node next()
+        {
+            Node n = next;
+            while (n == null)
+            {
+                // TODO: Thread.onSpinWait
+                Thread.yield();
```

Comment:
Looking at the Vyukov MpscLinkedQueue poll we tends to not use Thread::yield although I understand that with a single threaded machine could be the right way to handle the "bubble" while awaiting next to materialize: I would left a raw spin wait here